### PR TITLE
Nightly Packs Install

### PR DESCRIPTION
**Pull request by Shellyber**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/29250

## Minimum version of Demisto
- [x] 6.0.0

## Does it break backward compatibility?
   - [x] No

### CODE REVIEW

**Comment by Shellyber** [comment]
Example of failed nightly build that reruns and succeeded: https://app.circleci.com/pipelines/github/demisto/content/38992/workflows/94e7c643-b5cb-40bb-91d3-9e0a40b03c64/jobs/168669

File: Tests/Marketplace/search_and_install_packs.py [hunk 1]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+        host (str): The server URL.
+        prints_manager (ParallelPrintsManager): Print manager object.
+        thread_index (int): the thread index.
+        packs_to_install (list): A list of the packs to install.
+        request_timeout (int): Timeout settings for the installation request.
+
+    Returns:
+
+    """
+    packs_to_install_str = ', '.join([pack['id'] for pack in packs_to_install])
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:185]:
```suggestion
    packs_to_install_str = '\n '.join([pack['id'] for pack in packs_to_install])
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 2]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+    while PACK_INSTALL is not True:
+        try:
+            response_data, status_code, _ = demisto_client.generic_request_func(client,
+                                                                                path='/contentpacks/marketplace/install',
+                                                                                method='POST',
+                                                                                body=request_data,
+                                                                                accept='application/json',
+                                                                                _request_timeout=request_timeout)
+
+            if 200 <= status_code < 300:
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:205]: Add a debug log with the response.
The response in this case is a list of dicts with the keys: **id**, **name**, **lastInstallDate**, **updateAvailable**.
Example response:
```
[{'id': 'rasterize', 'currentVersion': '1.0.4', 'name': 'Rasterize', 'lastInstallDate': '2020-10-12T22:13:45.85429826Z', 'updateAvailable': False}]
```

I think we should print the name and the id for each pack

File: Tests/Marketplace/search_and_install_packs.py [hunk 3]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+                if line.startswith('HTTP response body: '):
+                    error_message = json.loads(line.split(': ', 1)[1])
+                    error = error_message.get('error')
+                    pack_id = error.split()[-2]
+            # Removed the bad pack from the list
+            packs = [pack for pack in packs_to_install if not (pack['id'] == pack_id)]
+            request_data = {
+                'packs': packs,
+                'ignoreWarnings': True
+            }
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:236]: I think this logic should be in it's own method where you can document the voodoo that's being done here and unittest it.

File: Tests/Marketplace/search_and_install_packs.py [hunk 4]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+
+    Returns:
+
+    """
+    packs_to_install_str = ', '.join([pack['id'] for pack in packs_to_install])
+    message = 'Installing the following packs in server {}:\n{}'.format(host, packs_to_install_str)
+    prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN, include_timestamp=True)
+    prints_manager.execute_thread_prints(thread_index)
+    # make the pack installation request
+    global PACK_INSTALL
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:190]: Why is it global if you're using it only in this method?

Comment (hod-alpert) [reply Tests/Marketplace/search_and_install_packs.py:190]: Also, I'd consider renaming it to something more indicative like ALL_PACKS_INSTALLED_SUCCESSFULY or something like that

File: Tests/Marketplace/search_and_install_packs.py [hunk 5]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+    PACK_INSTALL = False
+    request_data = {
+        'packs': packs_to_install,
+        'ignoreWarnings': True
+    }
+    while PACK_INSTALL is not True:
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:196]:
```suggestion
    while not PACK_INSTALL:
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 6]
```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+
+        except Exception as e:
+            err_msg = f'The request to install packs has failed. Reason:\n{str(e)}\n'
+            prints_manager.add_print_job(err_msg, print_error, thread_index, include_timestamp=True)
+            PACK_INSTALL = False
+            pack_id = ''
+            message = str(e).split('\n')
+            # Get the pack ID of the failed pack.
+            for line in message:
+                if line.startswith('HTTP response body: '):
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:227]: If this line does not exist you can enter infinite loop
Why not use regex instead?

File: Tests/Marketplace/search_and_install_packs.py [hunk 7]
```diff
@@ -224,43 +295,48 @@ def install_packs(client, host, prints_manager, thread_index, packs_to_install,
...
+                response_data, status_code, _ = demisto_client.generic_request_func(client,
+                                                                                    path='/contentpacks/marketplace/install',
+                                                                                    method='POST',
+                                                                                    body=request_data,
+                                                                                    accept='application/json',
+                                                                                    _request_timeout=request_timeout)
+
+                if 200 <= status_code < 300:
+                    message = 'Packs were successfully installed!\n'
+                    prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN,
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:324]: While you're in it please print the response here too.

And use `pformat` to print the tables (see example number 4 in the screenshot attached.

image

**Review by hod-alpert — CHANGES_REQUESTED** [review]

File: Tests/Marketplace/search_and_install_packs.py (hunk 1 above)
Comment (Shellyber) [reply Tests/Marketplace/search_and_install_packs.py]: Its better to leave the ', '. This way the print is not too long

File: Tests/Marketplace/search_and_install_packs.py (hunk 7 above)
Comment (Shellyber) [reply Tests/Marketplace/search_and_install_packs.py:350]: Sure about this? @hod-alpert 
It looks really bad, since we have 10 instances running, its making a lot of prints, lots on noise

**Comment by Shellyber** [comment]
@hod-alpert All changes are done. Please review again.
You can look at the last build to see all the prints and that all is good.
https://app.circleci.com/pipelines/github/demisto/content/39357/workflows/b49e0026-a0ea-4a43-8e98-85ab13a5e842/jobs/169910

File: Packs/AutoFocus/ReleaseNotes/1_2_0.md [hunk 8]
```diff
@@ -0,0 +1 @@
+This is test
```
Comment (hod-alpert) [inline Packs/AutoFocus/ReleaseNotes/1_2_0.md:1]: Revert these changes

File: Packs/AutoFocus/pack_metadata.json [hunk 9]
```diff
@@ -2,7 +2,7 @@
     "name": "AutoFocus",
     "description": "Use the Palo Alto Networks AutoFocus integration to distinguish the most\n  important threats from everyday commodity attacks.",
     "support": "xsoar",
-    "currentVersion": "1.1.9",
+    "currentVersion": "1.2.0",
```
Comment (hod-alpert) [inline Packs/AutoFocus/pack_metadata.json:5]: Revert these changes

File: Tests/Marketplace/search_and_install_packs.py [hunk 10]
```diff
@@ -168,8 +169,98 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+                                                                                method='POST',
+                                                                                body=request_data,
+                                                                                accept='application/json',
+                                                                                _request_timeout=request_timeout)
+            packs_data = []
+            if 200 <= status_code < 300:
+                message = 'Packs were successfully installed!\n'
+                prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN,
+                                             include_timestamp=True)
+                for pack in ast.literal_eval(response_data):
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:228]:
```suggestion
                for pack in response_data:
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 11]
```diff
@@ -168,8 +169,98 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+    }
+    while not all_packs_install_successfully:
+        try:
+            response_data, status_code, _ = demisto_client.generic_request_func(client,
+                                                                                path='/contentpacks/marketplace/install',
+                                                                                method='POST',
+                                                                                body=request_data,
+                                                                                accept='application/json',
+                                                                                _request_timeout=request_timeout)
+            packs_data = []
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:223]:
```suggestion
            packs_data = [{'ID': pack.get('id'),'CurrentVersion': pack.get('currentVersion')} for pack in response_data]
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 12]
```diff
@@ -168,8 +169,98 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+            if 200 <= status_code < 300:
+                message = 'Packs were successfully installed!\n'
+                prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN,
+                                             include_timestamp=True)
+                for pack in ast.literal_eval(response_data):
+                    packs_data.append({
+                        'ID': pack.get('id'),
+                        'CurrentVersion': pack.get('currentVersion')
+                    })
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:232]:
```suggestion

```

File: Tests/Marketplace/search_and_install_packs.py [hunk 13]
```diff
@@ -168,8 +169,98 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+
+    Returns: Pack_id (str)
+
+    """
+    malformed_pack_pattern = re.compile(r'invalid version [0-9.]+ for pack with ID ([\w_-]+)')
+    malformed_pack_id = malformed_pack_pattern.findall(str(error_message))
+    if malformed_pack_id:
+        return malformed_pack_id
+    else:
+        raise Exception(f'The request to install packs has failed. Reason: {str(error_message)}')
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:186]: Raising exception here will not break the while loop.
You should handle the case where you don't find a pack outside this method and break the loop

File: Tests/Marketplace/search_and_install_packs.py [hunk 14]
```diff
@@ -168,8 +169,98 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+        try:
+            response_data, status_code, _ = demisto_client.generic_request_func(client,
+                                                                                path='/contentpacks/marketplace/install',
+                                                                                method='POST',
+                                                                                body=request_data,
+                                                                                accept='application/json',
+                                                                                _request_timeout=request_timeout)
+            packs_data = []
+            if 200 <= status_code < 300:
+                message = 'Packs were successfully installed!\n'
```
Comment (hod-alpert) [inline Tests/Marketplace/search_and_install_packs.py:225]: Remove this print, you have the one with the packs data already

File: Tests/Marketplace/search_and_install_packs.py (hunk 7 above)
Comment (hod-alpert) [reply Tests/Marketplace/search_and_install_packs.py:344]: Yes yes, it's an important print, I'll make it look better in the future

**Review by hod-alpert — CHANGES_REQUESTED** [review]

**Review by hod-alpert — APPROVED** [review]

File: Tests/Marketplace/Tests/search_and_install_packs_test.py [hunk 15]
```diff
@@ -211,3 +211,24 @@ def test_search_pack_with_failure(mocker):
...
+S-Expiration=MDIgTm92IDIwIDExOjIzICswMDAw; Path=/; Expires=Mon, 02 Nov 2020 11:23:10 GMT; Max-Age=3600;
+Secure; SameSite=Lax', 'Strict-Transport-Security': 'max-age=10886400000000000; includeSubDomains',
+'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'DENY', 'X-Xss-Protection': '1; mode=block',
+'Date': 'Mon, 02 Nov 2020 10:23:10 GMT', 'Content-Length': '218'})
+HTTP response body: {"id":"bad_request","status":400,"title":"Bad request","detail":"Request body is not well-formed.
+It must be JSON.","error":"invalid version 1.2.0 for pack with ID AutoFocus (35000)","encrypted":false,"multires":null}
+"""
+
+
+def test_find_malformed_pack_id():
```
Comment (Itay4) [inline Tests/Marketplace/Tests/search_and_install_packs_test.py:232]: add test docstring

File: Tests/Marketplace/search_and_install_packs.py [hunk 16]
```diff
@@ -168,8 +169,94 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
         lock.release()
 
 
+def find_malformed_pack_id(error_message: str):
```
Comment (Itay4) [inline Tests/Marketplace/search_and_install_packs.py:172]:
```suggestion
def find_malformed_pack_id(error_message: str) -> str:
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 17]
```diff
@@ -168,8 +169,94 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+    corrupted pack.
+    Args:
+        client(demisto_client): The configured client to use.
+        host (str): The server URL.
+        prints_manager (ParallelPrintsManager): Print manager object.
+        thread_index (int): the thread index.
+        packs_to_install (list): A list of the packs to install.
+        request_timeout (int): Timeout settings for the installation request.
+
+    Returns:
```
Comment (Itay4) [inline Tests/Marketplace/search_and_install_packs.py:202]:
```suggestion
    Returns:
        None: No data returned.
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 18]
```diff
@@ -168,8 +169,94 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+    """
+    malformed_pack_pattern = re.compile(r'invalid version [0-9.]+ for pack with ID ([\w_-]+)')
+    malformed_pack_id = malformed_pack_pattern.findall(str(error_message))
+    if malformed_pack_id:
+        return malformed_pack_id
+    else:
+        raise Exception(f'The request to install packs has failed. Reason: {str(error_message)}')
+
+
+def install_nightly_packs(client, host, prints_manager, thread_index, packs_to_install, request_timeout=999999):
```
Comment (Itay4) [inline Tests/Marketplace/search_and_install_packs.py:189]: add type hints

File: Tests/Marketplace/search_and_install_packs.py (hunk 16 above)
Comment (Shellyber) [reply Tests/Marketplace/search_and_install_packs.py:172]: It returns list.
Will add that

File: Tests/Marketplace/search_and_install_packs.py [hunk 19]
```diff
@@ -168,8 +169,94 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+                'ignoreWarnings': True
+            }
+
+        finally:
+            prints_manager.execute_thread_prints(thread_index)
+
+
 def install_packs(client, host, prints_manager, thread_index, packs_to_install, request_timeout=999999,
-                  private_install=False):
+                  private_install=False, is_nightly=False):
```
Comment (Itay4) [inline Tests/Marketplace/search_and_install_packs.py:259]: add `is_nightly` to the docstring

File: Tests/Marketplace/search_and_install_packs.py [hunk 20]
```diff
@@ -168,8 +169,94 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
...
+                prints_manager.add_print_job(packs_message, print_color, thread_index, LOG_COLORS.GREEN,
+                                             include_timestamp=True)
+            else:
+                result_object = ast.literal_eval(response_data)
+                message = result_object.get('message', '')
+                err_msg = f'Failed to install packs - with status code {status_code}\n{message}\n'
+                prints_manager.add_print_job(err_msg, print_error, thread_index, include_timestamp=True)
+                raise Exception(err_msg)
+
+            all_packs_install_successfully = True
```
Comment (Itay4) [inline Tests/Marketplace/search_and_install_packs.py:237]: you can remove this as you `break` right after it anyway

File: Tests/Marketplace/Tests/search_and_install_packs_test.py [hunk 21]
```diff
@@ -211,3 +211,32 @@ def test_search_pack_with_failure(mocker):
...
+    """
+   Given
+   - Error message.
+   When
+   - Run find_malformed_pack_id command.
+   Then
+   - Ensure the pack ID is caught.
+   """
+    malformed_pack_id = script.find_malformed_pack_id(ERROR_MESSAGE)
```
Comment (Itay4) [inline Tests/Marketplace/Tests/search_and_install_packs_test.py:241]: lets also add case for no malformed id found
